src/wallet/tx_builder.rs

```diff
@@ -361,69 +391,57 @@ impl<D: Database, Cs: CoinSelectionAlgorithm<D>, Ctx: TxBuilderContext> TxBuilde
     ///
     /// This is useful for offline signers that take part to a multisig. Some hardware wallets like
     /// BitBox and ColdCard are known to require this.
-    pub fn add_global_xpubs(mut self) -> Self {
-        self.add_global_xpubs = true;
+    pub fn add_global_xpubs(&mut self) -> &mut Self {
+        self.params().add_global_xpubs = true;
         self
     }
 
     /// Spend all the available inputs. This respects filters like [`TxBuilder::unspendable`] and the change policy.
-    pub fn drain_wallet(mut self) -> Self {
-        self.drain_wallet = true;
+    pub fn drain_wallet(&mut self) -> &mut Self {
+        self.params().drain_wallet = true;
         self
     }
 
     /// Choose the coin selection algorithm
     ///
     /// Overrides the [`DefaultCoinSelectionAlgorithm`](super::coin_selection::DefaultCoinSelectionAlgorithm).
+    ///
+    /// Note that this function consumes the builder and returns it so it is usually best to put this as the first call on the builder.
```

Comment:
Is this note saying `coin_selection` (if you're overriding the default) should be set first due to performance reasons or something else? I noticed that `coin_selection` isn't used until the `finalize` function is called so seems whenever it is added to the builder (as long as before `finalize`) shouldn't matter, or am I missing something?

Reply:
Also does it make sense to check and panic if `coin_selection` is `None` as you do with `params`?

Reply:
> Is this note saying coin_selection (if you're overriding the default) should be set first due to performance reasons or something else?

No it's because of ergonomics:

```rust
// this doesn't work
wallet.build_tx().add_recipient(addr.script_pubkey(), 50_000).coin_selection(my_algorithm);
// this works
wallet.build_tx().coin_selection(my_algorithm).add_recipient(addr.script_pubkey(), 50_000);
```

The reason is because coin selection is a type parameter and this means we cannot do `self.coin_selection = new_coin_selection` (like you can with the other methods) because changing the `coin_selection` will also change the type of `Self`. So `coin_selection` has to consume `self` and return a new type of `Self`. But the other methods return `&mut self` which cannot be consumed. Therefore, you have to fix the type up front by doing `coin_selection` first.

There are a couple ways of improving this situation:

1. Don't pass in the database to the coin selection function on the trait. This would allow us to remove the type parameter on the `CoinSelection` trait and just have a `&dyn CoinSelection` in `TxParams`. IMO this would be fine even for those people who want to do database lookups during coin selection; they can just hold a reference to their database in an `Arc` or something. I guess no generic coin selection algorithm needs the database.
2. Just make `coin_selection` an argument to `.finish()` or `build_tx()` and `build_fee_bump()` so there is no ambiguity in the order.

My preference would be for (1) or leave it as it is.  @afilini thoughts?

> Also does it make sense to check and panic if coin_selection is None as you do with params?

Yes I think so! c234be1

Reply:
Thanks for the explanation, I just wanted to understand it better and that helps. It'd be a bit more user friendly if `.coin_selection` didn't have to go first but for now I'd say it's OK as is and if needed can tackle improvements in a future PR.

Reply:
See my answer as part of https://github.com/bitcoindevkit/bdk/pull/258#issuecomment-754685962

Reply:
This issue has been resolved.